File: 12.带最小值操作的栈/Solution.py
class MinStack(object):

    def __init__(self):
        # do some intialize if necessary
        self.data = list()
        self.dataM = list()

    def push(self, number):
        # write yout code here
        if len(self.data) == 0 or number <= self.dataM[-1]:
            self.dataM.append(number)
        else:
            self.dataM.append(self.dataM[-1])
        self.data.append(number)
            

    def pop(self):
        # pop and return the top item in stack
        self.dataM.pop()
        return self.data.pop()

    def min(self):
        # return the minimum number in stack
        return self.dataM[-1]

File: 12.带最小值操作的栈/Solution.cpp
class MinStack {
public:
    MinStack() {
        // do initialization if necessary
        
    }

    void push(int number) {
        // write your code here
        data.push(number);
        if(mins.empty() || number <= mins.top())
            mins.push(number);
        else{
            mins.push(mins.top());
        }
    }

    int pop() {
        // write your code here
        mins.pop();
        int res = data.top();
        data.pop();
        return res;
    }

    int min() {
        // write your code here
        int res = mins.top();
    }

private:
    stack<int> data;
    stack<int> mins;
};
